agent/http.go

```diff
@@ -83,15 +87,82 @@ type HTTPServer struct {
 	// proto is filled by the agent to "http" or "https".
 	proto string
 }
+type templatedFile struct {
+	templated *bytes.Reader
+	name      string
+	mode      os.FileMode
+	modTime   time.Time
+}
+
+func newTemplatedFile(buf *bytes.Buffer, raw http.File) *templatedFile {
+	info, _ := raw.Stat()
+	return &templatedFile{
+		templated: bytes.NewReader(buf.Bytes()),
+		name:      info.Name(),
+		mode:      info.Mode(),
+		modTime:   info.ModTime(),
+	}
+}
+
+func (t *templatedFile) Read(p []byte) (n int, err error) {
+	return t.templated.Read(p)
+}
+
+func (t *templatedFile) Seek(offset int64, whence int) (int64, error) {
+	return t.templated.Seek(offset, whence)
+}
+
+func (t *templatedFile) Close() error {
+	return nil
+}
+
+func (t *templatedFile) Readdir(count int) ([]os.FileInfo, error) {
+	return nil, errors.New("not a directory")
+}
+
+func (t *templatedFile) Stat() (os.FileInfo, error) {
+	return t, nil
+}
+
+func (t *templatedFile) Name() string {
+	return t.name
+}
+
+func (t *templatedFile) Size() int64 {
+	return int64(t.templated.Len())
+}
+
+func (t *templatedFile) Mode() os.FileMode {
+	return t.mode
+}
+
+func (t *templatedFile) ModTime() time.Time {
+	return t.modTime
+}
+
+func (t *templatedFile) IsDir() bool {
+	return false
+}
+
+func (t *templatedFile) Sys() interface{} {
+	return nil
+}
 
 type redirectFS struct {
-	fs http.FileSystem
+	fs          http.FileSystem
+	ContentPath string
 }
 
 func (fs *redirectFS) Open(name string) (http.File, error) {
 	file, err := fs.fs.Open(name)
```

Comment:
Should the index.html name check come before this Open, so that you don't double-open?

Reply:
Do you mean like:
```
	if name == "/index.html" || err != nil {
		file, err = fs.fs.Open("/index.html")
		content, _ := ioutil.ReadAll(file)
		file.Seek(0, 0)
		t, _ := template.New("fmtedindex").Parse(string(content))
		var out bytes.Buffer
		err = t.Execute(&out, fs)
		file = newTemplatedFile(&out, file)
	}else{
        file, err := fs.fs.Open(name)
}
return file,err
}
```

_Excuse the bad formatting_

Reply:
We should prevent double open but I think I proposed a better way overall down in one of my comments.

Create a `templatedIndexFS` type which all it does is Opens the file and if its name was index.html templates it. The redirectFS would wrap that and make the logic of each Open much cleaner.